SC/collectives: add change logs

minsii opened the pull request.

minsii committed:
section/collectives: reorder reduction op for active-set colls

The reduction operators mentioned in the intro of active-set-based
collectives did not follow a specific order, i.e., neither alphabetized
nor grouped by operator types. This patch reorders it to match the
description of team-based collectives. The new order also matches that
of the operator definitions in section 9.11.7 SHMEM_REDUCTIONS.

minsii committed:
section/collectives: fix sync example description

minsii committed:
section/collectives: fix typo in sync_all description

minsii committed:
Adding change log for bcast root update

minsii committed:
section/collectives: fix bcast example description

minsii committed:
section/collectives: fix typo in alltoall API description

minsii committed:
Adding change log for team-based collectives

minsii commented:
This PR contains several fixes and change logs for the collective section. @nspark @jdinan @davidozog Please review.

jdinan commented on content/backmatter.tex:
```diff
@@ -522,6 +522,25 @@ \section{Version 1.5}
 %
+\item Changed the broadcast function to update the dest object on all PEs,
+including the root PE.
+\\ See Section \ref{subsec:shmem_broadcast}.
```
This change is only for team-based broadcasts, not active-set based broadcasts. Can you please clarify?

jdinan commented on content/backmatter.tex:
```diff
@@ -522,6 +522,25 @@ \section{Version 1.5}
+\\ See Section \ref{subsec:shmem_broadcast}.
+%
+\item Clarified error codes returned from collectives.
+\\ See Section \ref{subsec:coll}.
```
I believe that the active-set based collectives did not return error codes. Since this only applies to the new collectives, it seems unnecessary to note this as a separate changelog entry.

minsii replied:
Will delete the change log. However, the main text does not specify that the error codes are team-based collectives specific. How about we change the title?
```
Error codes returned from collectives
```
->
```
Error codes returned from team-based collectives
```

jdinan replied:
Did we update the legacy collectives to also return error codes? I think this was discussed but ultimately not incorporated.

minsii replied:
We didn't. Same for `shmem_barrier_all` and `shmem_sync_all`.  Do you think we should leave the paragraph title as is?

jdinan replied:
A clarification is something we would make to existing text. Given that the error codes are newly added in OpenSHMEM 1.5, this changelog entry seems confusing to me. I think it would be fine to remove it and assume that this change is covered by another entry indicating the new team-based collectives. However, if you feel like it's helpful to call out this change in particular, it could be phrased a bit more like a new addition:

> Specified that \openshmem defines only success and failure error codes for team-based collective routines.

minsii replied:
Sorry for my late reply. I agree that the error code change is covered by the new team-based collectives entry. Thus, I deleted the change log entry and updated the paragraph title to: "Error codes returned from team-based collectives"

jdinan commented on content/backmatter.tex:
```diff
@@ -522,6 +522,25 @@ \section{Version 1.5}
+%
+\item Deprecated entire active-set-based collectives. Added the
+  team-based collective functions with return value to indicate successful
+  local completion: \FUNC{shmem\_sync},
```
Can we split this up?
```
\item Deprecated active-set-based collectives.
\item Added team-based collective functions:
```
The return value doesn't seem to me like a detail that needs to be called out here?

jdinan commented on content/shmem_sync_all.tex:
```diff
@@ -1,6 +1,6 @@
 \apisummary{
     Registers the arrival of a \ac{PE} at a synchronization point and suspends
-    execution until all other \acp{PE} in the default team arrive at a synchronization point. For multithreaded programs, execution is suspended
+    execution until all other \acp{PE} in the default team arrive at the synchronization point. For multithreaded programs, execution is suspended
```
Can we replace "default team" with `SHMEM_TEAM_WORLD` or "world team"?

minsii replied:
"default team" is used in several sections including teams, collectives, and memory management. It is defined in section 7. Library Handles.

I agree that `SHMEM_TEAM_WORLD` is more specific. Perhaps we want to replace it everywhere for consistency. Do we need to create separate commits for each SC? Is there a way that we can fix it in a single commit?

nspark replied:
I'd prefer "world team"—since, as with contexts, one could copy the handle to `SHMEM_TEAM_WORLD` and still refer to the same team—but this would be a "whole document" change (which I'd support).

minsii replied:
Good point on "world team". I created issue #371 to track the changes. I'd deal with the updates in separate commits and submit to the corresponding SC.

jdinan replied:
@nspark I'm fine with "world team" but we need to be precise about what that means. If one thread is calling `shmem_malloc`, is it a violation of the threading rules for another to call `shmem_team_barrier(SHMEM_TEAM_WORLD)` at the same time?

minsii replied:
Thinking again about the example @nspark mentioned: one does `handle = SHMEM_TEAM_WORLD;` and uses `handle` in later calls. The value of `handle` is still the predefined `SHMEM_TEAM_WORLD`, and the function call reads the value but not `&handle`.  Thus, I don't think people may confuse if the `handle` is a different team.

However, if we say "world team", I am afraid that people may confuse if a duplicated team of `SHMEM_TEAM_WORLD` (e.g., created by `shmem_team_split_strided`) is also a "world team".

In summary, I am now thinking `SHMEM_TEAM_WORLD` might be a better replacement for "default team." Does it resolve @jdinan's question, or avoid such confusion?

jdinan replied:
In Section 7, you can clarify that `SHMEM_TEAM_WORLD` handle corresponds to the "world team". Currently it says:
> Handle of type shmem_team_t that corresponds to the default team of all PEs in the OpenSHMEM program.

minsii replied:
I am working on the default team->world team replacement now. It will be in separate PRs.

nspark commented on content/collective_intro.tex:
```diff
@@ -55,7 +55,7 @@ \subsubsection*{Team-based collectives}
 \item \FUNC{shmem\_\{TYPE\_\}collect\{mem\}}
 \item \FUNC{shmem\_\{TYPE\_\}fcollect\{mem\}}
-\item Reduction routines for the following operations: AND, OR, XOR, MAX, MIN, SUM, PROD
+\item Reduction routines for the following operations: AND, MAX, MIN, SUM, PROD, OR, XOR
```
Why the reordering? The new ordering isn't alphabetized, which would be: AND, MAX, MIN, OR, PROD, SUM, XOR. The old order was, as I saw it, ordered as {bitwise, comparison, arithmetic} operators.

minsii replied:
I reordered it to match with the description for active-set based collectives. It looks weird to me that they describe very similar contents but in different orders. Maybe it is better to use "AND, OR, XOR, MAX, MIN, SUM, PROD" for both.  Suggestion?

nspark replied:
Yes, we should reorder the line for the active-set based collectives to match that of the team-based collectives:

https://github.com/nspark/specification/blob/14e66bed2fceda1567310f9e2cb4b5bf15056250/content/collective_intro.tex#L115

This will match the ordering of the routine listings in the Specification itself.

minsii replied:
Fixed.

nspark commented on content/shmem_broadcast.tex:
```diff
@@ -145,7 +145,7 @@
 \apicexample
     {In the following \Cstd[11] example, the call to \FUNC{shmem\_broadcast} copies \source{}
-    on \ac{PE} $0$ to \dest{} on \acp{PE} $1\dots npes-1$.
+    on \ac{PE} $0$ to \dest{} on \acp{PE} $0\dots npes-1$.
```
Good catch!

jdinan commented:
You could also specify in the description of the SHMEM_TEAM_WORLD handle
that it corresponds to the "world team".

On Wed, Feb 12, 2020, 12:22 PM Min Si <notifications@github.com> wrote:

> *@minsii* commented on this pull request.
> ------------------------------
>
> In content/shmem_sync_all.tex
> <https://github.com/nspark/specification/pull/10#discussion_r378398059>:
>
> > @@ -1,6 +1,6 @@
>  \apisummary{
>      Registers the arrival of a \ac{PE} at a synchronization point and suspends
> -    execution until all other \acp{PE} in the default team arrive at a synchronization point. For multithreaded programs, execution is suspended
> +    execution until all other \acp{PE} in the default team arrive at the synchronization point. For multithreaded programs, execution is suspended
>
> Thinking again about the example @nspark <https://github.com/nspark>
> mentioned: one does handle = SHMEM_TEAM_WORLD; and uses handle in later
> calls. The value of handle is still the predefined SHMEM_TEAM_WORLD, and
> the function call reads the value but not &handle. Thus, I don't think
> people may confuse if the handle is a different team.
>
> However, if we say "world team", I am afraid that people may confuse if a
> duplicated team of SHMEM_TEAM_WORLD (e.g., created by
> shmem_team_split_strided) is also a "world team".
>
> In summary, I am now thinking SHMEM_TEAM_WORLD might be a better
> replacement for "default team." Does it resolve @jdinan
> <https://github.com/jdinan>'s question, or avoid such confusion?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nspark/specification/pull/10?email_source=notifications&email_token=AAZ5P4WAYL572NHLQFQ56I3RCQV3XA5CNFSM4KRYGQSKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCVISYCY#discussion_r378398059>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAZ5P4S4WOHJXWRPFT75NUTRCQV3XANCNFSM4KRYGQSA>
> .
>

minsii committed:
section/collectives: clarify err code is only for team based routines

minsii commented:
@nspark @jdinan @davidozog I think all comments should be fixed except the "default team"->"world team" replacement. I will address it in separate PRs. This PR should be ready for review & merge. Could you please check again?

nspark commented on content/backmatter.tex:
```diff
@@ -522,6 +522,23 @@ \section{Version 1.5}
 \\ See Section \ref{subsec:shmmallochint} and \ref{subsec:library_constants}.
 %
+\item Changed the team-based broadcast function to update the dest object on
+all PEs, including the root PE.
```
```suggestion
all \acp{PE}, including the root \ac{PE}.
```

nspark commented on content/backmatter.tex:
```diff
@@ -522,6 +522,23 @@ \section{Version 1.5}
 \CONST{SHMEM\_MALLOC\_ATOMICS\_REMOTE} and \CONST{SHMEM\_MALLOC\_SIGNAL\_REMOTE}.
 \\ See Section \ref{subsec:shmmallochint} and \ref{subsec:library_constants}.
 %
+\item Changed the team-based broadcast function to update the dest object on
```
```suggestion
\item Changed the team-based broadcast function to update the \VAR{dest} object on
```

nspark reviewed (changes requested):
Looks good; only requesting two minor formatting changes

nspark committed:
Apply suggestions from code review

Minor formatting tweaks

nspark commented on content/backmatter.tex:
```diff
@@ -522,6 +522,23 @@ \section{Version 1.5}
 \CONST{SHMEM\_MALLOC\_ATOMICS\_REMOTE} and \CONST{SHMEM\_MALLOC\_SIGNAL\_REMOTE}.
 \\ See Section \ref{subsec:shmmallochint} and \ref{subsec:library_constants}.
 %
+\item Changed the team-based broadcast function to update the \VAR{dest} object on
```
```suggestion
\item Specified that team-based broadcast operations update the \VAR{dest} object on
```

nspark committed:
Update changelog wording for team-based broadcasts

Minor rewording, as team-based broadcasts are new, so we're not changing behavior, but it's worth calling out here that the behavior is different from that of active-set-based broadcasts.

nspark committed:
Merge branch 'section/collectives' into pr/sc/collectives/changelogs
